smart-metastore/src/main/java/org/smartdata/metastore/dao/accesscount/AccessCountTableManager.java

```diff
@@ -151,44 +153,198 @@ private void recoverTables() {
     }
   }
 
-  private List<AccessCountTable> getTablesForLast(
-      long startTime, long endTime) throws MetaStoreException {
+  private List<AccessCountTable> getAccessCountTables(long startTime,
+                                                      long endTime) throws MetaStoreException {
+    return getAccessCountTables(startTime, endTime, null);
+  }
 
-    List<AccessCountTable> results = new ArrayList<>();
-    TimeGranularity searchIntervalGranularity = TimeGranularity.of(endTime - startTime);
+  private List<AccessCountTable> getAccessCountTables(long startTime,
+                                                      long endTime,
+                                                      AccessCountTable parentTable)
+      throws MetaStoreException {
+    List<AccessCountTable> result = new ArrayList<>();
+    final TimeGranularity startGranularity = TimeGranularity.of(endTime - startTime);
+    TimeGranularity searchIntervalGranularity = startGranularity;
+    boolean decreaseOnly = false;
     do {
-      // Here we assume that tables are sorted by time.
-      for (AccessCountTable table : inMemoryTableManager.getTablesOfGranularity(
-          searchIntervalGranularity)) {
-
-        // skip interval if it ends before specified search start time
-        if (table.getEndTime() <= startTime) {
+      long finalStartTime = startTime;
+      Collection<AccessCountTable> tables = inMemoryTableManager.getTablesOfGranularity(
+              searchIntervalGranularity).stream()
+          .filter(t -> t.getStartTime() <= endTime)
+          .collect(Collectors.toList());
+      boolean foundTables = !tables.isEmpty();
+      if (parentTable == null && !decreaseOnly) {
+        //when we didn't find any tables of current granularity,
+        // we have to try to find tables of bigger granularity
+        while (!foundTables) {
+          searchIntervalGranularity = increaseGranularity(searchIntervalGranularity);
+          if (searchIntervalGranularity == null) {
+            tables = Collections.emptyList();
+            searchIntervalGranularity = startGranularity;
+            foundTables = true;
+          } else {
+            tables = inMemoryTableManager.getTablesOfGranularity(
+                searchIntervalGranularity);
+            Long accessCountTableStartTime = tables.stream()
+                .findFirst()
+                .map(AccessCountTable::getStartTime)
+                .orElse(Long.MAX_VALUE);
+            foundTables = accessCountTableStartTime < startTime;
```

Comment:
Why not `accessCountTableStartTime <= startTime`?

Reply:
fixed